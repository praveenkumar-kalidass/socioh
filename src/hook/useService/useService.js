import * as Keychain from 'react-native-keychain';
import AsyncStorage from '@react-native-community/async-storage';
import RNContacts from 'react-native-contacts';

import { CONSTANT, ERROR, TRANSLATION } from '../../constant';
import useAjax from '../useAjax';
import useUser from '../useUser';

const useService = () => {
  const { ajax } = useAjax();
  const { user, saveUserDetails, clearUserDetails } = useUser();

  const signUp = async ({ name, email, password }) => {
    await Keychain.resetGenericPassword();
    await Keychain.setGenericPassword(email, password);
    await AsyncStorage.setItem(
      CONSTANT.ASYNC_STORAGE_KEY.SOCIOH_USER,
      JSON.stringify({
        name,
        email,
      }),
    );
    saveUserDetails({
      name,
      email,
    });
    return ajax();
  };

  const signIn = async ({ email, password }) => {
    const credentials = await Keychain.getGenericPassword();
    if (credentials.username === email && credentials.password === password) {
      const userDetails = await AsyncStorage.getItem(
        CONSTANT.ASYNC_STORAGE_KEY.SOCIOH_USER,
      );
      saveUserDetails(JSON.parse(userDetails));
      return ajax();
    }
    throw Error(ERROR.INVALID_CREDENTIALS);
  };

  const logout = async () => {
    clearUserDetails();
    return ajax();
  };

  const getUserDetails = async () => {
    if (user) {
      await ajax();
      return {
        ...user,
        interests: Object.values(TRANSLATION.INTERESTS),
      };
    }
    throw Error(ERROR.NO_USER_FOUND);
  };

  const getContacts = async (search) => {
    const permission = await RNContacts.requestPermission();
    if (permission === CONSTANT.PERMISSION.AUTHORIZED) {
      let contacts;
      if (search) {
        contacts = await RNContacts.getContactsMatchingString(search);
      } else {
        contacts = await RNContacts.getAll();
      }
      await ajax();
      return contacts;
    }
    throw Error(ERROR.UNAUTHORIZED_TO_READ_CONTACTS);
  };

  const getMessages = async () => {
    await ajax();
    return CONSTANT.MESSAGES.GREETINGS;
  };

  const getMessageList = async () => {
    await ajax();
    return CONSTANT.MESSAGES.LIST;
  };

  return {
    signUp,
    signIn,
    logout,
    getUserDetails,
    getContacts,
    getMessages,
    getMessageList,
  };
};

export default useService;
