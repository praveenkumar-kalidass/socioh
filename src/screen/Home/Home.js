import React from 'react';
import PropTypes from 'prop-types';
import { Thumbnail } from 'native-base';

import User from '../../asset/image/user.png';
import useUser from '../../hook/useUser';
import { ICON, ROUTE } from '../../constant';
import {
  AvatarContainer,
  AvatarLabel,
  HomeContainer,
  HomeLayout,
  UserNavigationItem,
  UserNavigationList,
  NavigationLabel,
  NavigationIcon,
} from './Home.style';

const Home = ({ navigation }) => {
  const { user } = useUser();

  const navigateToProfile = () => {
    navigation.navigate(ROUTE.PROFILE);
  };

  return (
    <HomeContainer>
      <HomeLayout>
        <AvatarContainer testID="home_avatar" onPress={navigateToProfile}>
          <Thumbnail large source={User} />
          <AvatarLabel bold>{user.name}</AvatarLabel>
        </AvatarContainer>
        <UserNavigationList>
          <UserNavigationItem>
            <NavigationIcon type={ICON.MATERIAL_FAMILY} name={ICON.HOME} />
            <NavigationLabel>Home</NavigationLabel>
          </UserNavigationItem>
        </UserNavigationList>
      </HomeLayout>
    </HomeContainer>
  );
};

Home.propTypes = {
  navigation: PropTypes.shape({
    navigate: PropTypes.func,
  }).isRequired,
};

export default Home;
