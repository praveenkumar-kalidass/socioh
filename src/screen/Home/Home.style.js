import styled from 'styled-components/native';
import { Text, List, ListItem, Icon } from 'native-base';

import { COLOR, FONT, SPACING } from '../../constant';

export const HomeContainer = styled.View`
  background-color: ${COLOR.PALE_GREEN};
  flex: 1;
`;

export const HomeLayout = styled.View`
  background-color: ${COLOR.PRIMARY};
  border-top-left-radius: ${SPACING.SPACE_20};
  border-top-right-radius: ${SPACING.SPACE_20};
  flex: 1;
  margin-top: 30%;
`;

export const AvatarContainer = styled.View`
  align-items: center;
  padding-vertical: ${SPACING.SPACE_20};
`;

export const AvatarLabel = styled(Text)`
  color: ${COLOR.WHITE};
  font-size: ${FONT.SIZE_24};
  margin-top: ${SPACING.SPACE_10};
`;

export const UserNavigationList = styled(List)``;

export const UserNavigationItem = styled(ListItem)`
  margin-left: ${SPACING.SPACE_0};
  padding-left: ${SPACING.SPACE_20};
`;

export const NavigationIcon = styled(Icon)`
  color: ${COLOR.WHITE};
`;

export const NavigationLabel = styled(Text)`
  color: ${COLOR.WHITE};
  font-size: ${FONT.SIZE_16};
  font-weight: bold;
  margin-left: ${SPACING.SPACE_20};
  text-transform: uppercase;
`;
