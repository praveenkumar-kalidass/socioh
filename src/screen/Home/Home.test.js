import React from 'react';
import { render, act, fireEvent, waitFor } from '@testing-library/react-native';

import useService from '../../hook/useService';
import Home from './index';

const serviceMock = {
  getUserDetails: jest.fn(),
  logout: jest.fn(),
};

jest.mock('../../hook/useService');

describe('Home', () => {
  const mockNavigation = { navigate: jest.fn() };

  beforeEach(() => {
    serviceMock.getUserDetails.mockResolvedValueOnce({ name: 'Praveen' });
    useService.mockImplementation(() => serviceMock);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot', () => {
    const container = render(<Home />);

    expect(container).toMatchSnapshot();
  });

  it('should load user details', async () => {
    const { getByTestId } = render(<Home navigation={mockNavigation} />);

    await waitFor(() => getByTestId('home_user_name'));

    expect(getByTestId('home_user_name').props.children).toStrictEqual(
      'Praveen',
    );
  });

  it('should not show user details, when service throws error', async () => {
    serviceMock.getUserDetails.mockRejectedValueOnce(Error('NO_USER_FOUND'));

    const { queryByTestId } = render(<Home navigation={mockNavigation} />);

    expect(queryByTestId('home_user_name')).toBeNull();
  });

  it('should navigate to profile screen when avatar is pressed', () => {
    const { getByTestId } = render(<Home navigation={mockNavigation} />);

    act(() => {
      fireEvent.press(getByTestId('home_avatar'));
    });

    expect(mockNavigation.navigate).toHaveBeenCalledTimes(1);
    expect(mockNavigation.navigate).toHaveBeenCalledWith('PROFILE');
  });

  it('should call logout from service, when logout link is pressed', async () => {
    serviceMock.logout.mockResolvedValueOnce();

    const { getByTestId } = render(<Home navigation={mockNavigation} />);

    await waitFor(() => getByTestId('home_user_name'));

    await act(async () => {
      await fireEvent.press(getByTestId('home_logout'));
    });

    expect(mockNavigation.navigate).toHaveBeenCalledTimes(1);
    expect(mockNavigation.navigate).toHaveBeenCalledWith('LOGIN');
  });

  it('should navigate to messages screen when messages link is pressed', () => {
    const { getByTestId } = render(<Home navigation={mockNavigation} />);

    act(() => {
      fireEvent.press(getByTestId('home_messages'));
    });

    expect(mockNavigation.navigate).toHaveBeenCalledTimes(1);
    expect(mockNavigation.navigate).toHaveBeenCalledWith('MESSAGES');
  });
});
