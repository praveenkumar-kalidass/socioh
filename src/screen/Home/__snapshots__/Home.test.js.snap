// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#b0e6e7",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#1abfbf",
          "borderTopLeftRadius": 20,
          "borderTopRightRadius": 20,
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "marginTop": "30%",
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "opacity": 1,
          "paddingVertical": 20,
        }
      }
      testID="home_avatar"
    >
      <Image
        large={true}
        source={
          Object {
            "testUri": "../../../src/asset/image/user.png",
          }
        }
        style={
          Object {
            "borderRadius": 40,
            "height": 80,
            "width": 80,
          }
        }
      />
    </View>
    <View
      style={
        Array [
          Object {},
          Object {},
        ]
      }
    >
      <View
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View
          selected={true}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderBottomWidth": 0.5,
                "borderColor": "#c9c9c9",
                "flexDirection": "row",
                "marginLeft": 16,
                "paddingBottom": 13,
                "paddingRight": 16,
                "paddingTop": 13,
              },
              Object {
                "backgroundColor": "#b0e6e7",
                "marginLeft": 0,
                "paddingLeft": 20,
              },
            ]
          }
        >
          <Text
            allowFontScaling={false}
            selected={true}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 12,
                },
                Array [
                  Object {
                    "color": "#000",
                    "fontSize": 30,
                  },
                  Object {
                    "color": "#007aff",
                    "fontSize": 30,
                  },
                  Object {
                    "color": "#000000",
                  },
                ],
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            type="MaterialIcons"
          >
            
          </Text>
          <Text
            selected={true}
            style={
              Array [
                Object {
                  "alignSelf": "center",
                  "color": "#007aff",
                  "fontFamily": "System",
                  "fontSize": 16,
                },
                Object {
                  "color": "#000000",
                  "fontSize": 16,
                  "fontWeight": "bold",
                  "marginLeft": 20,
                  "textTransform": "uppercase",
                },
              ]
            }
          >
            Home
          </Text>
        </View>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="home_contacts"
      >
        <View
          onPress={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderBottomWidth": 0.5,
                "borderColor": "#c9c9c9",
                "flexDirection": "row",
                "marginLeft": 16,
                "paddingBottom": 13,
                "paddingRight": 16,
                "paddingTop": 13,
              },
              Object {
                "marginLeft": 0,
                "paddingLeft": 20,
              },
            ]
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 12,
                },
                Array [
                  Object {
                    "color": "#000",
                    "fontSize": 30,
                  },
                  Object {
                    "color": "#000",
                    "fontSize": 30,
                  },
                  Object {
                    "color": "#ffffff",
                  },
                ],
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            type="MaterialIcons"
          >
            
          </Text>
          <Text
            style={
              Array [
                Object {
                  "alignSelf": "center",
                  "color": "#000",
                  "fontFamily": "System",
                  "fontSize": 16,
                },
                Object {
                  "color": "#ffffff",
                  "fontSize": 16,
                  "fontWeight": "bold",
                  "marginLeft": 20,
                  "textTransform": "uppercase",
                },
              ]
            }
          >
            Contacts
          </Text>
        </View>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="home_messages"
      >
        <View
          onPress={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderBottomWidth": 0.5,
                "borderColor": "#c9c9c9",
                "flexDirection": "row",
                "marginLeft": 16,
                "paddingBottom": 13,
                "paddingRight": 16,
                "paddingTop": 13,
              },
              Object {
                "marginLeft": 0,
                "paddingLeft": 20,
              },
            ]
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 12,
                },
                Array [
                  Object {
                    "color": "#000",
                    "fontSize": 30,
                  },
                  Object {
                    "color": "#000",
                    "fontSize": 30,
                  },
                  Object {
                    "color": "#ffffff",
                  },
                ],
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            type="MaterialIcons"
          >
            
          </Text>
          <Text
            style={
              Array [
                Object {
                  "alignSelf": "center",
                  "color": "#000",
                  "fontFamily": "System",
                  "fontSize": 16,
                },
                Object {
                  "color": "#ffffff",
                  "fontSize": 16,
                  "fontWeight": "bold",
                  "marginLeft": 20,
                  "textTransform": "uppercase",
                },
              ]
            }
          >
            Messages
          </Text>
        </View>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="home_contacts"
      >
        <View
          onPress={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderBottomWidth": 0.5,
                "borderColor": "#c9c9c9",
                "flexDirection": "row",
                "marginLeft": 16,
                "paddingBottom": 13,
                "paddingRight": 16,
                "paddingTop": 13,
              },
              Object {
                "marginLeft": 0,
                "paddingLeft": 20,
              },
            ]
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 12,
                },
                Array [
                  Object {
                    "color": "#000",
                    "fontSize": 30,
                  },
                  Object {
                    "color": "#000",
                    "fontSize": 30,
                  },
                  Object {
                    "color": "#ffffff",
                  },
                ],
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            type="MaterialIcons"
          >
            
          </Text>
          <Text
            style={
              Array [
                Object {
                  "alignSelf": "center",
                  "color": "#000",
                  "fontFamily": "System",
                  "fontSize": 16,
                },
                Object {
                  "color": "#ffffff",
                  "fontSize": 16,
                  "fontWeight": "bold",
                  "marginLeft": 20,
                  "textTransform": "uppercase",
                },
              ]
            }
          >
            Photos
          </Text>
        </View>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="home_logout"
      >
        <View
          button={true}
          onPress={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderBottomWidth": 0.5,
                "borderColor": "#c9c9c9",
                "flexDirection": "row",
                "marginLeft": 16,
                "paddingBottom": 13,
                "paddingRight": 16,
                "paddingTop": 13,
              },
              Object {
                "marginLeft": 0,
                "paddingLeft": 20,
              },
            ]
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 12,
                },
                Array [
                  Object {
                    "color": "#000",
                    "fontSize": 30,
                  },
                  Object {
                    "color": "#000",
                    "fontSize": 30,
                  },
                  Object {
                    "color": "#ffffff",
                  },
                ],
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            type="MaterialIcons"
          >
            
          </Text>
          <Text
            style={
              Array [
                Object {
                  "alignSelf": "center",
                  "color": "#000",
                  "fontFamily": "System",
                  "fontSize": 16,
                },
                Object {
                  "color": "#ffffff",
                  "fontSize": 16,
                  "fontWeight": "bold",
                  "marginLeft": 20,
                  "textTransform": "uppercase",
                },
              ]
            }
          >
            Logout
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
